pkg/controllers/metrics/node/controller.go

```diff
@@ -0,0 +1,340 @@
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Register(ctx context.Context, m manager.Manager) error {
+	err := controllerruntime.
+		NewControllerManagedBy(m).
+		Named("nodemetrics").
+		For(&v1.Node{}).
+		Watches(
+			// Reconcile all nodes related to a provisioner when it changes.
+			&source.Kind{Type: &v1alpha5.Provisioner{}},
+			handler.EnqueueRequestsFromMapFunc(func(o client.Object) (requests []reconcile.Request) {
+				nodes := &v1.NodeList{}
+				if err := c.KubeClient.List(ctx, nodes, client.MatchingLabels(map[string]string{v1alpha5.ProvisionerNameLabelKey: o.GetName()})); err != nil {
+					logging.FromContext(ctx).Debugf("Failed to list nodes when mapping expiration watch events, %s", err.Error())
+					return requests
+				}
+				for _, node := range nodes.Items {
+					requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{Name: node.Name}})
+				}
+				return requests
+			}),
+		).
+		Watches(
+			// Reconcile node when a pod assigned to it changes.
+			&source.Kind{Type: &v1.Pod{}},
+			handler.EnqueueRequestsFromMapFunc(func(o client.Object) (requests []reconcile.Request) {
+				if name := o.(*v1.Pod).Spec.NodeName; name != "" {
+					requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{Name: name}})
+				}
+				return requests
+			}),
+		).
+		Complete(c)
+	return err
+}
+
+func (c *Controller) deleteGauges(labelSlice []prometheus.Labels) {
+	for _, labels := range labelSlice {
+		allocatableGaugeVec.Delete(labels)
+		requestsGaugeVec.Delete(labels)
+		limitsGaugeVec.Delete(labels)
+		overheadGaugeVec.Delete(labels)
+	}
+}
+
+// generateLabels creates the labels using the current state of the pod
+func (c *Controller) generateLabels(node *v1.Node, resourceTypeName string) (prometheus.Labels, error) {
+	metricLabels := prometheus.Labels{}
+	metricLabels[resourceType] = resourceTypeName
+	metricLabels[nodeName] = node.GetName()
+	metricLabels[nodeProvisioner] = node.Labels[v1alpha5.ProvisionerNameLabelKey]
+	metricLabels[nodeZone] = node.Labels[v1.LabelTopologyZone]
+	metricLabels[nodeArchitecture] = node.Labels[v1.LabelArchStable]
+	metricLabels[nodeCapacityType] = node.Labels[v1alpha5.LabelCapacityType]
+	metricLabels[nodeInstanceType] = node.Labels[v1.LabelInstanceTypeStable]
+	metricLabels[nodePhase] = string(node.Status.Phase)
+	// Add node labels
+	labels, err := json.Marshal(node.GetLabels())
+	if err != nil {
+		return nil, fmt.Errorf("marshal pod labels: %w", err)
+	}
+	metricLabels[nodeLabels] = string(labels)
+	return metricLabels, nil
+}
+
+func (c *Controller) updateGauges(ctx context.Context, node *v1.Node) error {
+
+	// Calculate total requests and limits from all pods
+	podlist := &v1.PodList{}
+	if err := c.KubeClient.List(ctx, podlist, client.MatchingFields{"spec.nodeName": node.Name}); err != nil {
+		return fmt.Errorf("listing pods on node %s, %w", node.Name, err)
+	}
+	reqs, limits := getPodsTotalRequestsAndLimits(podlist.Items)
```

Comment:
Nit:
I'd use `requests` to avoid introducing a new noun. I'm all for simplifying/aliasing when things are overly verbose, but I think it muddies things rather than clarifies them.

Reply:
yes agree! done